### PR TITLE
Fixed #25989 -- Find latest lastmod of all sitemaps

### PR DESCRIPTION
**Pull request by allixx**
Determine latest lastmod of all sitemaps supplied to
contrib.sitemaps.views.sitemap, instead of unconditionally picking
lastmod of a last sitemap.

### CODE REVIEW

File: django/contrib/sitemaps/views.py [hunk 1]
```diff
@@ -65,16 +66,23 @@ def sitemap(request, sitemaps, section=None,
...
         except EmptyPage:
             raise Http404("Page %s empty" % page)
         except PageNotAnInteger:
             raise Http404("No page '%s'" % page)
     response = TemplateResponse(request, template_name, {'urlset': urls},
                                 content_type=content_type)
-    if hasattr(site, 'latest_lastmod'):
-        # if latest_lastmod is defined for site, set header so as
+    if lastmod is not None:
+        # if lastmod is defined for at least one site, set header so as
```
Comment (charettes) [inline django/contrib/sitemaps/views.py]: Shouldn't we only set `'Last-Modified'` to the most recent `lastmod` if **all** sites define one and don't set anything if at least one site doesn't define one?

Comment (allixx) [inline django/contrib/sitemaps/views.py]: I have to agree with you here. The code is rewritten to explicitly require having lastmod defined for all sitemaps.

File: django/contrib/sitemaps/views.py [hunk 2]
```diff
@@ -65,16 +67,26 @@ def sitemap(request, sitemaps, section=None,
...
                                       protocol=req_protocol))
+            if all_sites_lastmod:
+                site_lastmod = getattr(site, 'latest_lastmod', None)
+                if site_lastmod is not None:
+                    if type(site_lastmod) is datetime.date:
+                        site_lastmod = datetime.datetime(site_lastmod.year,
+                                                         site_lastmod.month,
+                                                         site_lastmod.day)
+                    if lastmod is None or lastmod < site_lastmod:
+                        lastmod = site_lastmod
```
Comment (charettes) [inline django/contrib/sitemaps/views.py]: `max()` handles the `None` case correctly.

File: django/contrib/sitemaps/views.py [hunk 3]
```diff
@@ -65,16 +67,26 @@ def sitemap(request, sitemaps, section=None,
                 site = site()
             urls.extend(site.get_urls(page=page, site=req_site,
                                       protocol=req_protocol))
+            if all_sites_lastmod:
+                site_lastmod = getattr(site, 'latest_lastmod', None)
+                if site_lastmod is not None:
+                    if type(site_lastmod) is datetime.date:
```
Comment (charettes) [inline django/contrib/sitemaps/views.py]: Use `isinstance`

File: django/contrib/sitemaps/views.py [hunk 4]
```diff
@@ -65,16 +67,26 @@ def sitemap(request, sitemaps, section=None,
...
+                        site_lastmod = datetime.datetime(site_lastmod.year,
+                                                         site_lastmod.month,
+                                                         site_lastmod.day)
```
Comment (charettes) [inline django/contrib/sitemaps/views.py]: Maybe you could convert `datetime` to `utctimetuple()` and `date` to `timetuple()` here instead. That's the actual type you'll need for `timegm` anyway and you won't have to deal with naive and aware datetime comparison this way.

**Comment by allixx** [comment]
Tests added.

**Comment by charettes** [comment]
Buildbot, test this please.

**Comment by allixx** [comment]
Fixed python35 `TypeError: unorderable types: time.struct_time() > NoneType()` error.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/sitemaps_tests/test_http.py [hunk 5]
```diff
@@ -110,6 +110,38 @@ def test_sitemap_last_modified_mixed(self):
         response = self.client.get('/lastmod-mixed/sitemap.xml')
         self.assertFalse(response.has_header('Last-Modified'))
 
+    def test_sitemaps_lastmod_mixed_ascending_last_modified_missing(self):
+        """
+        Tests that Last-Modified header is omitted when lastmod not found in
```
Comment (xblitz) [inline tests/sitemaps_tests/test_http.py]: Why would the  Last-Modified header not be set when it is not found in ALL maps?   shouldn't it be set as soon as one is found.. and only NOT set if no last_mod is found.

Comment (allixx) [inline tests/sitemaps_tests/test_http.py]: Docs don't explicitly define this behaviour. If we stretch lastmod calculation logic of a `Sitemap` class to this case, all sitemaps should have a lasmod.

Comment (xblitz) [inline tests/sitemaps_tests/test_http.py]: Ah yes.. I see what you mean.

**Comment by timgraham** [comment]
merged in 002a4f72c46458514d97f7e54d4fd9416d89899d, thanks!

**Pull request closed without merging**